asserts/preseed.go

```diff
@@ -0,0 +1,218 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"crypto"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+type PreseedSnap struct {
+	Name     string
+	SnapID   string
+	Revision int
+}
+
+// SnapName implements naming.SnapRef.
+func (s *PreseedSnap) SnapName() string {
+	return s.Name
+}
+
+// ID implements naming.SnapRef.
+func (s *PreseedSnap) ID() string {
+	return s.SnapID
+}
+
+type Preseed struct {
+	assertionBase
+	snaps     []*PreseedSnap
+	timestamp time.Time
+}
+
+func (p *Preseed) Series() string {
+	return p.HeaderString("series")
+}
+
+// AccountID returns the identifier of the account that signed this assertion.
+func (p *Preseed) AccountID() string {
+	return p.HeaderString("account-id")
+}
+
+// BrandID returns the brand identifier. Same as the authority id.
+func (p *Preseed) BrandID() string {
+	return p.HeaderString("brand-id")
+}
+
+// Model returns the model name identifier.
+func (p *Preseed) Model() string {
+	return p.HeaderString("model")
+}
+
+// SystemLabel returns the label of the seeded system.
+func (p *Preseed) SystemLabel() string {
+	return p.HeaderString("system-label")
+}
+
+// Timestamp returns the time when the preseed assertion was issued.
+func (p *Preseed) Timestamp() time.Time {
+	return p.timestamp
+}
+
+// PreseedSHA3_384 returns the checksum of preseeding artifact.
+func (p *Preseed) PreseedSHA3_384() string {
+	return p.HeaderString("preseed-sha3-384")
+}
+
+// Snaps returns the snaps for preseeding
+func (p *Preseed) Snaps() []*PreseedSnap {
+	return p.snaps
+}
+
+func checkPreseedSnap(snap map[string]interface{}) (*PreseedSnap, error) {
+	name, err := checkNotEmptyStringWhat(snap, "name", "of snap")
+	if err != nil {
+		return nil, err
+	}
+	if err := naming.ValidateSnap(name); err != nil {
+		return nil, fmt.Errorf("invalid snap name %q", name)
+	}
+
+	what := fmt.Sprintf("of snap %q", name)
+
+	var snapID string
```

Comment:
probably worth a comment that snap-id can be omitted if the model allows for unasserted snaps